cpp/src/io/parquet/stream_pool.cpp

```diff
@@ -0,0 +1,148 @@
+
+namespace {
+
+// TODO: what is a good number here. what's the penalty for making it larger?
+// Dave Baranec rule of thumb was max_streams_needed * num_concurrent_threads,
+// where num_concurrent_threads was estimated to be 4. so using 32 will allow
+// for 8 streams per thread, which should be plenty (decoding will be up to 4
+// kernels when delta_byte_array decoding is added). rmm::cuda_stream_pool
+// defaults to 16.
+std::size_t constexpr STREAM_POOL_SIZE = 32;
+
+class rmm_cuda_stream_pool : public cuda_stream_pool {
+  rmm::cuda_stream_pool _pool;
+
+ public:
+  rmm_cuda_stream_pool() : _pool{STREAM_POOL_SIZE} {}
+  rmm::cuda_stream_view get_stream() override { return _pool.get_stream(); }
+  rmm::cuda_stream_view get_stream(std::size_t stream_id) override
+  {
+    return _pool.get_stream(stream_id);
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) override
+  {
+    static std::mutex stream_pool_mutex;
+
+    if (count > STREAM_POOL_SIZE) {
+      CUDF_LOG_WARN("get_streams called with count ({}) > pool size ({})", count, STREAM_POOL_SIZE);
+    }
+    auto streams = std::vector<rmm::cuda_stream_view>();
+    std::lock_guard<std::mutex> lock(stream_pool_mutex);
+    for (uint32_t i = 0; i < count; i++) {
+      streams.emplace_back(_pool.get_stream());
+    }
+    return streams;
+  }
+
+  std::size_t get_stream_pool_size() const override { return STREAM_POOL_SIZE; }
+};
+
+class debug_cuda_stream_pool : public cuda_stream_pool {
+ public:
+  rmm::cuda_stream_view get_stream() override { return cudf::get_default_stream(); }
+  rmm::cuda_stream_view get_stream(std::size_t stream_id) override
+  {
+    return cudf::get_default_stream();
+  }
+
+  std::vector<rmm::cuda_stream_view> get_streams(uint32_t count) override
+  {
+    return std::vector<rmm::cuda_stream_view>(count, cudf::get_default_stream());
+  }
+
+  std::size_t get_stream_pool_size() const override { return 1UL; }
+};
+
+cuda_stream_pool* create_global_cuda_stream_pool()
+{
+  if (getenv("LIBCUDF_USE_DEBUG_STREAM_POOL")) return new debug_cuda_stream_pool();
+
+  return new rmm_cuda_stream_pool();
+}
+
+// implementation of per-thread-default-event.
+class cuda_event_map {
+ public:
+  cuda_event_map() {}
+
+  cudaEvent_t find(std::thread::id thread_id)
+  {
+    std::lock_guard<std::mutex> lock(map_mutex_);
+    auto it = event_map_.find(thread_id);
+    if (it != event_map_.end()) {
+      return it->second;
+    } else {
+      cudaEvent_t event;
+      CUDF_CUDA_TRY(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+      event_map_[thread_id] = event;
+      return event;
+    }
+  }
+
+  cuda_event_map(cuda_event_map const&) = delete;
+  void operator=(cuda_event_map const&) = delete;
+
+ private:
+  std::unordered_map<std::thread::id, cudaEvent_t> event_map_;
+  std::mutex map_mutex_;
+};
+
+cudaEvent_t event_for_thread()
+{
+  static cuda_event_map instance;
+  return instance.find(std::this_thread::get_id());
+}
```

Comment:
No need for the `cuda_event_map`. `thread_local` will handle creating an object per thread for you. 

```suggestion
struct cuda_event {
  cuda_event()
      : e_{[]() {
          cudaEvent_t event;
          CUDF_CUDA_TRY(
              cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
          return event;
        }()} {}

 private:
  struct deleter {
    using pointer = cudaEvent_t;
    auto operator()(cudaEvent_t e) { cudaEventDestroy(e); }
  };
  std::unique_ptr<cudaEvent_t, deleter> e_;
};


cudaEvent_t event_for_thread()
{
  thread_local cuda_event  thread_event;
  return thread_event;
}
```